File: Examples/ore-rest-api/src/main/java/com/ore/api/service/OREWorker.java
package com.ore.api.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.ore.api.config.OREConfig;
import com.ore.api.dto.OREInputs;
import com.ore.api.model.Job;
import com.ore.api.repository.JobRepository;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;

/**
 * Service for executing ORE subprocess and parsing results.
 */
@Slf4j
@Service
public class OREWorker {

    private final OREConfig oreConfig;
    private final ObjectMapper objectMapper;
    private final JobRepository jobRepository;
    private final StorageService storageService;
    private final OREResultParser resultParser;
    private final Path jobsBaseDir;

    public OREWorker(
            OREConfig oreConfig,
            ObjectMapper objectMapper,
            JobRepository jobRepository,
            StorageService storageService,
            OREResultParser resultParser,
            @Value("${ore.jobs.base-dir}") String jobsBaseDir
    ) {
        this.oreConfig = oreConfig;
        this.objectMapper = objectMapper;
        this.jobRepository = jobRepository;
        this.storageService = storageService;
        this.resultParser = resultParser;
        this.jobsBaseDir = Path.of(jobsBaseDir);
    }

    /**
     * Execute ORE asynchronously with @Async support.
     */
    @org.springframework.scheduling.annotation.Async("oreTaskExecutor")
    public CompletableFuture<Void> executeORE(String jobId) {
        return CompletableFuture.runAsync(() -> {
            try {
                doExecute(jobId);
            } catch (Exception e) {
                log.error("Error executing ORE job {}", jobId, e);
                markJobFailed(jobId, e.getMessage());
            }
        });
    }

    private void doExecute(String jobId) throws Exception {
        log.info("Starting ORE execution for job {}", jobId);

        // Update status to running
        updateStatus(jobId, "running", 10);

        Path inputDir = storageService.getJobInputDir(jobId);
        Path outputDir = storageService.getJobOutputDir(jobId);

        // Check ORE executable
        if (!oreConfig.isOREAvailable()) {
            throw new IllegalStateException("ORE executable not found. Configure ore.executable.path or set ORE_PATH environment variable.");
        }

        Path oreExe = Path.of(oreConfig.getPath());
        Path oreXml = inputDir.resolve("ore.xml");

        if (!Files.exists(oreXml)) {
            throw new IllegalStateException("ore.xml not found in job input directory");
        }

        // Build process
        ProcessBuilder pb = new ProcessBuilder(
            oreExe.toString(),
            "ore.xml"
        );
        pb.directory(inputDir.toFile());
        pb.redirectErrorStream(true);

        log.debug("Executing ORE: {} in directory {}", oreExe, inputDir);

        Process process = pb.start();

        // Capture output with timeout
        StringBuilder output = new StringBuilder();
        try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8))
        ) {
            String line;
            while ((line = reader.readLine()) != null) {
                output.append(line).append("\n");
            }
        }

        // Wait for process with timeout
        boolean completed = process.waitFor(oreConfig.getTimeoutSeconds(), TimeUnit.SECONDS);

        if (!completed) {
            process.destroyForcibly();
            throw new RuntimeException(String.format(
                "ORE execution timed out after %d seconds", oreConfig.getTimeoutSeconds()
            ));
        }

        int exitCode = process.exitValue();

        // Save log to output directory
        Files.createDirectories(outputDir);
        Path logFile = outputDir.resolve("log.txt");
        Files.writeString(logFile, output.toString(), StandardCharsets.UTF_8);

        if (exitCode != 0) {
            String error = output.length() > 2000 ? output.substring(output.length() - 2000) : output.toString();
            throw new RuntimeException(String.format("ORE exited with code %d: %s", exitCode, error.trim()));
        }

        // Parse results
        Job job = jobRepository.findById(jobId).orElse(null);
        String jobType = job != null ? job.getJobType() : "unknown";
        Map<String, Object> summary = resultParser.parseSummary(outputDir, jobType);
        List<String> outputFiles = storageService.getOutputFiles(jobId);

        markJobCompleted(jobId, summary, outputFiles);

        log.info("Job {} completed successfully. Output files: {}", jobId, outputFiles);
    }

    private void updateStatus(String jobId, String status, int progress) {
        Job job = jobRepository.findById(jobId).orElse(null);
        if (job != null) {
            job.setStatus(status);
            job.setProgress(progress);
            jobRepository.save(job);
        }
    }

    private void markJobFailed(String jobId, String errorMessage) {
        Job job = jobRepository.findById(jobId).orElse(null);
        if (job != null) {
            job.setStatus("failed");
            job.setProgress(0);
            job.setErrorMessage(errorMessage != null && errorMessage.length() > 2000
                ? errorMessage.substring(errorMessage.length() - 2000)
                : errorMessage);
            jobRepository.save(job);
        }
    }

    private void markJobCompleted(String jobId, Map<String, Object> summary, List<String> files) {
        Job job = jobRepository.findById(jobId).orElse(null);
        if (job != null) {
            job.setStatus("completed");
            job.setProgress(100);
            try {
                job.setResultJson(objectMapper.writeValueAsString(Map.of(
                    "summary", summary,
                    "files", files
                )));
            } catch (Exception e) {
                log.warn("Failed to serialize result JSON for job {}", jobId, e);
            }
            jobRepository.save(job);
        }
    }
}

File: Examples/ORE-RESTAPI/app/workers/ore_worker.py
import json
import logging
import subprocess
from pathlib import Path

from app.config import settings
from app.database import SessionLocal
from app.services.result_parser import ResultParser
from app.services.storage_service import StorageService

logger = logging.getLogger(__name__)


def execute_ore_job(job_id: str) -> None:
    """Execute an ORE job as a subprocess.

    This function is designed to be called by RQ worker or synchronously.
    """
    db = SessionLocal()
    storage = StorageService()
    parser = ResultParser()

    try:
        from app.models.job import Job

        job = db.query(Job).filter(Job.id == job_id).first()
        if not job:
            logger.error("Job %s not found", job_id)
            return

        # Update status to running
        job.status = "running"
        job.progress = 10
        db.commit()

        input_dir = storage.get_job_input_dir(job_id)
        output_dir = storage.get_job_output_dir(job_id)
        ore_exe = str(_resolve_ore_exe())

        ore_xml_path = input_dir / "ore.xml"
        if not ore_xml_path.exists():
            raise FileNotFoundError(f"ore.xml not found in {input_dir}")

        logger.info("Starting ORE execution for job %s", job_id)

        # Execute ORE subprocess
        result = subprocess.run(
            [ore_exe, "ore.xml"],
            cwd=str(input_dir),
            capture_output=True,
            text=True,
            timeout=settings.ORE_TIMEOUT_SECONDS,
        )

        # Capture log output
        log_content = result.stdout
        if result.stderr:
            log_content += "\n" + result.stderr

        # Save log to output directory
        log_path = output_dir / "log.txt"
        log_path.parent.mkdir(parents=True, exist_ok=True)
        log_path.write_text(log_content, encoding="utf-8")

        # Check exit code
        if result.returncode != 0:
            error_msg = log_content.strip() or f"ORE exited with code {result.returncode}"
            job.status = "failed"
            job.error_message = error_msg[-2000:]  # Truncate long errors
            job.progress = 0
            logger.error("ORE execution failed for job %s: %s", job_id, error_msg[:200])
            db.commit()
            return

        # Parse results
        summary = parser.parse_summary(output_dir, job.job_type)
        output_files = storage.get_output_files(job_id)

        job.status = "completed"
        job.progress = 100
        job.result_json = json.dumps({
            "summary": summary,
            "files": output_files,
        })
        logger.info("Job %s completed. Output files: %s", job_id, output_files)
        db.commit()

    except subprocess.TimeoutExpired:
        _fail_job(db, job_id, f"ORE execution timed out after {settings.ORE_TIMEOUT_SECONDS}s")
        logger.error("ORE execution timed out for job %s", job_id)

    except Exception as e:
        _fail_job(db, job_id, str(e))
        logger.exception("Unexpected error executing job %s", job_id)

    finally:
        db.close()


def _resolve_ore_exe() -> Path:
    """Resolve the ORE executable path."""
    from app.config import resolve_ore_path
    return resolve_ore_path()


def _fail_job(db, job_id: str, error_message: str) -> None:
    """Mark a job as failed."""
    from app.models.job import Job

    job = db.query(Job).filter(Job.id == job_id).first()
    if job:
        job.status = "failed"
        job.error_message = error_message[-2000:]
        job.progress = 0
        db.commit()
